File: 2398-maximum-number-of-robots-within-budget/2398-maximum-number-of-robots-within-budget.py
class Solution:
    def maximumRobots(self, chargeTimes: List[int], runningCosts: List[int], budget: int) -> int:
        
        q = []
        f, l, currSum, ans, n = 0, 0, 0, 0, len(chargeTimes)
        
        while (f < n):
            while (len(q) != 0 and chargeTimes[q[-1]] < chargeTimes[f]): 
                q.pop()
            q.append(f)
            currSum += runningCosts[f]
            f += 1
            while (l != f and chargeTimes[q[0]] + (f-l) * currSum > budget):
                if q[0] == l:
                    q.pop(0)
                currSum -= runningCosts[l]
                l += 1
            ans = max(ans, f-l)
        
        return ans
            

File: 2398-maximum-number-of-robots-within-budget/2398-maximum-number-of-robots-within-budget.java
class Solution {
    public int maximumRobots(int[] chargeTimes, int[] runningCosts, long budget) {
        
        // monotonically decreasing queue
        Deque<Integer> q = new LinkedList<>();
        // window first and last pointers
        int n = chargeTimes.length, f = 0, l = 0, ans = 0;
        // window sum
        long currSum = 0;
        
        while (f < n) {
            // maintain the monotonically decreasing property of the queue
            while (q.size() > 0 && chargeTimes[q.peekLast()] < chargeTimes[f]) q.pollLast();
            q.addLast(f);
            currSum += runningCosts[f++];
            // adjust l pointer for the current f pointer
            while (f!=l && chargeTimes[q.peekFirst()]+(f-l)*currSum > budget) {
                if (q.peekFirst() == l) q.pollFirst();
                currSum -= runningCosts[l++];
            }
            ans = Math.max(ans, f-l);
        }
        
        return ans;
    }
}